pkg/storage/watch/watch.go

```diff
@@ -0,0 +1,196 @@
+package watch
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/rjeczalik/notify"
+	log "github.com/sirupsen/logrus"
+	"github.com/weaveworks/ignite/pkg/storage"
+	"github.com/weaveworks/ignite/pkg/storage/watch/update"
+)
+
+const eventBuffer = 4096 // How many events and updates we can buffer before watching is interrupted
+var excludeDirs = []string{".git"}
+var listenEvents = []notify.Event{notify.InCreate, notify.InDelete, notify.InDeleteSelf, notify.InCloseWrite}
+
+var eventMap = map[notify.Event]update.Event{
+	notify.InCreate:     update.EventCreate,
+	notify.InDelete:     update.EventDelete,
+	notify.InCloseWrite: update.EventModify,
+}
+
+type eventStream chan notify.EventInfo
+type UpdateStream chan *update.FileUpdate
+type watches []string
+
+type watcher struct {
+	dir          string
+	events       eventStream
+	updates      UpdateStream
+	watches      watches
+	suspendEvent update.Event
+	monitor      *Monitor
+}
+
+func (w *watcher) addWatch(path string) (err error) {
+	log.Tracef("Watcher: adding watch for %q", path)
+	if err = notify.Watch(path, w.events, listenEvents...); err == nil {
+		w.watches = append(w.watches, path)
+	}
+
+	return
+}
+
+func (w *watcher) hasWatch(path string) bool {
+	for _, watch := range w.watches {
+		if watch == path {
+			log.Tracef("Watcher: watch found for %q", path)
+			return true
+		}
+	}
+
+	log.Tracef("Watcher: no watch found for %q", path)
+	return false
+}
+
+func (w *watcher) clear() {
+	log.Tracef("Watcher: clearing all watches")
+	notify.Stop(w.events)
+	w.watches = w.watches[:0]
+}
+
+// newWatcher returns a list of files in the watched directory in
+// addition to the generated watcher, it can be used to populate
+// MappedRawStorage fileMappings
+func newWatcher(dir string) (w *watcher, files []string, err error) {
+	w = &watcher{
+		dir:     dir,
+		events:  make(eventStream, eventBuffer),
+		updates: make(UpdateStream, eventBuffer),
+	}
+
+	if err = w.start(&files); err != nil {
```

Comment:
I think returning `files, err := w.start()` is way clearer instead of the pointer pattern

Reply:
`w.start(nil)` is used inside the goroutine as well, by using the pointer I avoid performing a potentially expensive filename validation during `inotify` monitoring. This is why a pointer is used:
```golang
if files != nil {
	// Only include files with a valid suffix
	if validSuffix(info.Name()) {
		*files = append(*files, path)
	}
}
```